File: Searcha2DMatrixII240.java
/*
Time complexity : O(m+n)
Space complexity: O(1)
*/
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length ==0){
            return false;
        }
        int m = 0;
        int n = matrix[0].length-1;
        
        while(m < matrix.length && n>=0){
            if(matrix[m][n] == target){
                return true;
            }

            else if(target > matrix[m][n]){
                m++;
            }
            else{
                n--;
            }
        
        }
        return false;
    }
}

File: search-a-2d-matrix-ii_240.py
#Time Complexity O(MN) One pass for traversing matrix array using row column pointers
#Space Complexity O(1)
#It successfully runs on the leetcode

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix: return
        r= len(matrix)-1
        c= len(matrix[0])-1
        i = 0
        j = c
        while i<=r and j>=0:
            if matrix[i][j]==target: return True
            if matrix[i][j] > target:
                j-=1
            else:
                i+=1
        return False
        